### PR TITLE
Improved BatchRendererGroup API (for new Hybrid Renderer) and BatchRendererGroup Test Projects

### PR DESCRIPTION
**Pull request by cinight**
### IMPORTANT
See link in "What should be reviewed" section to find a filter that excludes the non-code files, which are the vast majority of the changed files in the PR

### Purpose of this PR
This PR contains the Graphics changes for the new BatchRendererGroup API which is also used by the new Hybrid Renderer with DOTS 1.0.
It is meant to work with the C++ part. The PR is still in flight here, and expected to land shortly: https://ono.unity3d.com/unity/unity/pull-request/134116/_/graphics/hrnext

Detailed technical design document here: https://docs.google.com/document/d/1WOj_Uaq-h9X0BRHZo1H5wJkYE-JxO8meE6Rv1_a6PjY/edit#

Draft user manual here: https://docs.google.com/document/d/10iAN_JUsTgjqdYUF-mtJzRvVK5IqsdVh8pBq5C25Vlo/edit#heading=h.5eo6s16tlfl3

### What should be reviewed
The PR changes are very large, but this is mainly because all the DOTS related projects that lived in this repo have been removed. Some new test projects that do not depend on DOTS have also been added. The size of the 'real' changes is much more manageable. 
You can find the filtered file query [here](https://github.com/Unity-Technologies/Graphics/pull/5842/files?authenticity_token=IpgI7HHJZvTN9rBr6UnyFAhLs7aGEddyiZ%2F%2BBD9geqnMAeCGDbcd%2BhynkUE5Lhobo1pJelobM5KYneqdda3vxg%3D%3D&file-filters%5B%5D=.cs&file-filters%5B%5D=.hlsl). Basically .cs and .hlsl changes.

### Technical changes
- Hybrid Renderer V1 has been completely removed, and ifdefs related to it have been removed. The name "Hybrid V2" is also removed. Going forwards, there will only be a single Hybrid Renderer (the third version, but not called "V3").
- Shader macros have been refactored to eliminate the workaround for a bug in the fxc preprocessor. The new macros are simpler and easier to use as a result, and will only work with the new caching preprocessor which is nowadays enabled by default.
- A number of built-in shader properties in DOTS instancing variants have been migrated from DOTS instanced properties to efficiently encoded constant buffer constants set up by dedicated SRP Batcher C++ code. This saves a significant amount of GPU memory and should also be faster.
- The DOTS instancing metadata constant naming format has been changed to avoid double underscores occurring. This fixes a problem on Nintendo Switch which emits compiler warnings when such names are encountered.
- The UnityPerDraw cbuffer is not defined anymore with the `DOTS_INSTANCING_ON` variant. BatchRendererGroup doesn't need it anymore and it's not even bound. 
- Fixed shader compilation errors related to previous matrix definitions with the `DOTS_INSTANCING_ON` variant. Hybrid Renderer had not support for Unity 2021 and the Hybrid Renderer/SRP CI has been disabled for the whole 2021 cycle. It comes at no surprise that the DOTS variant was broken at some point, so we fixed that.
- Enabled LOD fade variants for DOTS instanced HDRP Shader Graphs, for which they were disabled. BatchRendererGroup is now able to support it, and a future version of Hybrid Renderer will implement it (but does not currently do so).

### Test Projects changes
**UniversalHybridTest** and **HDRP_HybridTests** were originally for testing DOTS hybrid renderer.
But Hybrid Renderer test projects are now living in the DOTS repo. So in this PR we converted these 2 projects to be purely testing the new BRG APIs. The projects are now not depending on any DOTS packages.
These 2 projects are supposed to be automated, but the actual CI setup will be in future PRs as it need to wait until the C++ PR has landed. Therefore this PR only contains the project content changes.

---
### Testing status
[Test Doc](https://docs.google.com/document/d/1YeZwCRmBTuZ2A-EFUAuAS3T7ncJWYvMbHZ-dqf6sW38/edit#)

---
### Comments to reviewers
If you want to test, please refer to [How To Test](https://docs.google.com/document/d/1YeZwCRmBTuZ2A-EFUAuAS3T7ncJWYvMbHZ-dqf6sW38/edit#bookmark=kix.k5gkzl12eu5) section in TestDoc.

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/Material/Lit/LitProperties.hlsl [hunk 1]
```diff
@@ -293,13 +293,13 @@ float _TessellationObjectScale;
 float _TessellationTilingScale;
 #endif
 
+CBUFFER_END
+
```
Comment (vincent-breysse) [inline com.unity.render-pipelines.high-definition/Runtime/Material/Lit/LitProperties.hlsl:297]: TODO : Do this for all the other files including those IDs in the PerMaterial cbuffer

File: com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDTarget.cs [hunk 2]
```diff
@@ -846,39 +846,23 @@ static class CorePragmas
...
-        public static PragmaCollection DotsInstancedInV2Only = new PragmaCollection
+        public static PragmaCollection DotsInstanced = new PragmaCollection
         {
-            { Pragma.InstancingOptions(InstancingOptions.RenderingLayer) },
-            #if ENABLE_HYBRID_RENDERER_V2
+            { Basic },
+            { Pragma.MultiCompileInstancing },
             { Pragma.DOTSInstancing },
+            { Pragma.InstancingOptions(InstancingOptions.RenderingLayer) },
             { Pragma.InstancingOptions(InstancingOptions.NoLodFade) },
```
Comment (sebbbi) [inline com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/HDTarget.cs:855]: BatchRendererGroup API nowadays supports LOD fade. This line should be removed. Hybrid Renderer doesn't support LOD fade yet, but I would assume that the default value (zero) it feeds to the BRG API is OK. Hybrid Renderer also explicitly doesn't enable the DrawCommand LOD Crossfade bit, meaning that there should be no complications there.

**Review by sebbbi — APPROVED** [review]

**Review by JussiKnuuttila — APPROVED** [review]

**Review by joeldevahl — APPROVED** [review]

**Review by cdxntchou — CHANGES_REQUESTED** [review]
Love the changes!

Definitely make sure the shadergraph import version is bumped.

**Review by ernestasKupciunas — APPROVED** [review]
Looks good! I ran provided testing scenes using URP on some Android and iOS devices
Testing doc: https://docs.google.com/document/d/12xWskZtDc5aFrKH0BwIFPm3zzEq_t9LMye1K7VXR6Iw/edit

**Review by eh-unity — APPROVED** [review]
I looked at Universal side only, skimmed over the rest. LGTM.

**Review by cdxntchou — APPROVED** [review]
Looks good!

**Review by phi-lira — APPROVED** [review]

**Review by TomasKiniulis — APPROVED** [review]
All good on HDRP side. Approving on behalf HDRP QA. Test status in Test Doc of pr description.

**Comment by GrantLamb-Unity** [comment]
I ran some tests with the Hybrid Renderer/Shadergraph with the following setup:
`graphics/hrnext` C++ branch (Editor)
`hybrid-renderer-2021` DOTS branch
`graphics/hrnext` Graphics branch (this PR)

I verified that I was able to set up a per instance Material Override and I overrode a Color property of a Material with Shadergraph. I pretty much followed the steps [listed here](https://docs.unity3d.com/Packages/com.unity.rendering.hybrid@0.11/manual/material-overrides-asset.html). I'll be adding these steps and this workflow to our Shadergraph manual test suite in TestRail.~

**Review by GrantLamb-Unity — APPROVED** [review]

**Review by sebastienlagarde — APPROVED** [review]

**Review by arnaudcarre — APPROVED** [review]

**Pull request merged**